Source/Core/Logger/Logger.swift

```diff
@@ -0,0 +1,302 @@
+        return Double(nanoTime) / 1_000_000
+    }
+}
+
+/**
+ A logging class that can be told where to log to via transports.
+
+ Features include:
+ - Asynchronous, ordered output to transports
+ - All log messages can be tagged and filtered
+ - `LogLevel`s are provided as well
+ - Force logging enables output via `print` even if no transports available
+
+ ## Filtering
+
+ Two methods exist to allow for filtering of the log stream.
+
+ - `Logger.filterUnless(tag:)`
+ - `Logger.filterIf(tag:)`
+
+ */
+public class Logger {
+    /// Shared logger object
+    public static let shared: Logger = {
+        Logger()
+    }()
+
+    private let startTime = DispatchTime.now()
+    private let queue = DispatchQueue(label: "com.schibsted.identity.Logger", qos: .utility)
+    private let dispatchGroup = DispatchGroup()
+
+    private var _enabled: Bool = true
+    private var _outputTags: Bool = false
+    private var transports: [(String) -> Void] = []
+    private var allowedTags = Set<String>()
+    private var ignoredTags = Set<String>()
+    private let label: String
+
+    /**
+     Initializes a Logger object
+     */
+    public init(label: String? = nil) {
+        self.label = label ?? ""
+    }
+
+    /**
+     Transports are called asynchronously, if you need to wait till all logging output has been sent to
+     all transports then this function blocks until that happens
+     */
+    public func waitTillAllLogsTransported() {
+        self.dispatchGroup.wait()
+    }
+
+    /// Set to true if you want the tags to be printed as well
+    public var outputTags: Bool {
+        get {
+            return self.queue.sync {
+                self._outputTags
+            }
+        }
+        set {
+            self.queue.async { [weak self] in
+                self?._outputTags = newValue
+            }
+        }
+    }
+
+    /// If this is false then it ignores all logs
+    public var enabled: Bool {
+        get {
+            return self.queue.sync {
+                self._enabled
+            }
+        }
+        set {
+            self.queue.async { [weak self] in
+                self?._enabled = newValue
+            }
+        }
+    }
+
+    /**
+     Adding a transport allows you to tell the logger where the output goes to. You may add as
+     many as you like.
+
+     - parameter transport: function that is called with each log invocaton
+     */
+    public func addTransport(_ transport: @escaping (String) -> Void) {
+        self.queue.async { [weak self] in
+            self?.transports.append(transport)
+        }
+    }
+
+    /// Removes all transports
+    public func removeTransports() {
+        self.queue.async { [weak self] in
+            self?.transports.removeAll()
+        }
+    }
+
+    /// Filters log messages unless they are tagged with `tag`
+    public func filterUnless(tag: String) {
+        _ = self.queue.async { [weak self] in
```

Comment:
Some places you do `_ = self.queue.async()`other places you don't. Maybe do the same all places?